File: cocos/core/renderer/models/skinning-model.ts
/*
 Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
  worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You shall
  not use Cocos Creator software for developing other software or tools that's
  used for developing games. You are not granted to publish, distribute,
  sublicense, and/or sell copies of Cocos Creator.

 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

/**
 * @hidden
 */

import { EDITOR } from 'internal:constants';
import { Material } from '../../assets/material';
import { Mesh, RenderingSubMesh } from '../../assets/mesh';
import { Skeleton } from '../../assets/skeleton';
import { aabb } from '../../geometry';
import { GFXBuffer } from '../../gfx/buffer';
import { GFXBufferUsageBit, GFXMemoryUsageBit } from '../../gfx/define';
import { Mat4, Vec3 } from '../../math';
import { UBOSkinning } from '../../pipeline/define';
import { Node } from '../../scene-graph/node';
import { Pass, IMacroPatch } from '../core/pass';
import { ModelType } from '../scene/model';
import { uploadJointData } from './skeletal-animation-utils';
import { MorphModel } from './morph-model';
import { IPSOCreateInfo } from '../scene/submodel';

export interface IJointTransform {
    node: Node;
    local: Mat4;
    world: Mat4;
    stamp: number;
    parent: IJointTransform | null;
}

const stack: IJointTransform[] = [];
const pool: Map<string, IJointTransform> = new Map();

const myPatches = [
    { name: 'CC_USE_SKINNING', value: true },
];

export function getWorldMatrix (transform: IJointTransform | null, stamp: number) {
    let i = 0;
    let res = Mat4.IDENTITY;
    while (transform) {
        if (transform.stamp === stamp || transform.stamp + 1 === stamp && !transform.node.hasChangedFlags) {
            res = transform.world;
            transform.stamp = stamp;
            break;
        }
        transform.stamp = stamp;
        stack[i++] = transform;
        transform = transform.parent;
    }
    while (i > 0) {
        transform = stack[--i];
        const node = transform.node;
        Mat4.fromRTS(transform.local, node.rotation, node.position, node.scale);
        res = Mat4.multiply(transform.world, res, transform.local);
    }
    return res;
}

export function getTransform (node: Node, root: Node) {
    let joint: IJointTransform | null = null;
    let i = 0;
    while (node !== root) {
        const id = node.uuid;
        if (pool.has(id)) {
            joint = pool.get(id)!;
            break;
        } else { // TODO: object reuse
            joint = { node, local: new Mat4(), world: new Mat4(), stamp: -1, parent: null };
            pool.set(id, joint);
        }
        stack[i++] = joint;
        node = node.parent!;
        joint = null;
    }
    let child: IJointTransform;
    while (i > 0) {
        child = stack[--i];
        child.parent = joint;
        joint = child;
    }
    return joint;
}

export function deleteTransform (node: Node) {
    let transform = pool.get(node.uuid) || null;
    while (transform) {
        pool.delete(transform.node.uuid);
        transform = transform.parent;
    }
}

function getRelevantBuffers (outIndices: number[], outBuffers: number[], jointMaps: number[][], targetJoint: number) {
    for (let i = 0; i < jointMaps.length; i++) {
        const idxMap = jointMaps[i];
        let index = -1;
        for (let j = 0; j < idxMap.length; j++) {
            if (idxMap[j] === targetJoint) { index = j; break; }
        }
        if (index >= 0) {
            outBuffers.push(i);
            outIndices.push(index);
        }
    }
}

interface IJointInfo {
    bound: aabb;
    target: Node;
    bindpose: Mat4;
    transform: IJointTransform;
    buffers: number[];
    indices: number[];
}

const v3_min = new Vec3();
const v3_max = new Vec3();
const v3_1 = new Vec3();
const v3_2 = new Vec3();
const m4_1 = new Mat4();
const ab_1 = new aabb();

/**
 * @en
 * The skinning model that is using real-time pose calculation.
 * @zh
 * 实时计算动画的蒙皮模型。
 */
export class SkinningModel extends MorphModel {

    public uploadAnimation = null;

    private _buffers: GFXBuffer[] = [];
    private _dataArray: Float32Array[] = [];
    private _joints: IJointInfo[] = [];
    private _bufferIndices: number[] | null = null;

    constructor () {
        super();
        this.type = ModelType.SKINNING;
    }

    public destroy () {
        this.bindSkeleton();
        if (this._buffers.length) {
            for (let i = 0; i < this._buffers.length; i++) {
                this._buffers[i].destroy();
            }
            this._buffers.length = 0;
        }
        super.destroy();
    }

    public bindSkeleton (skeleton: Skeleton | null = null, skinningRoot: Node | null = null, mesh: Mesh | null = null) {
        for (let i = 0; i < this._joints.length; i++) {
            deleteTransform(this._joints[i].target);
        }
        this._bufferIndices = null; this._joints.length = 0;
        if (!skeleton || !skinningRoot || !mesh) { return; }
        this.transform = skinningRoot;
        const boneSpaceBounds = mesh.getBoneSpaceBounds(skeleton);
        const jointMaps = mesh.struct.jointMaps;
        this._ensureEnoughBuffers(jointMaps && jointMaps.length || 1);
        this._bufferIndices = mesh.jointBufferIndices;
        for (let index = 0; index < skeleton.joints.length; index++) {
            const bound = boneSpaceBounds[index];
            const target = skinningRoot.getChildByPath(skeleton.joints[index]);
            if (!bound || !target) { continue; }
            const transform = getTransform(target, skinningRoot)!;
            const bindpose = skeleton.bindposes[index];
            const indices: number[] = [];
            const buffers: number[] = [];
            if (!jointMaps) { indices.push(index); buffers.push(0); }
            else { getRelevantBuffers(indices, buffers, jointMaps, index); }
            this._joints.push({ indices, buffers, bound, target, bindpose, transform });
        }
    }

    public updateTransform (stamp: number) {
        const root = this.transform!;
        // @ts-ignore TS2445
        if (root.hasChangedFlags || root._dirtyFlags) {
            root.updateWorldTransform();
            this._transformUpdated = true;
        }
        // update bounds
        Vec3.set(v3_min,  Infinity,  Infinity,  Infinity);
        Vec3.set(v3_max, -Infinity, -Infinity, -Infinity);
        for (let i = 0; i < this._joints.length; i++) {
            const { bound, transform } = this._joints[i];
            const worldMatrix = getWorldMatrix(transform, stamp);
            aabb.transform(ab_1, bound, worldMatrix);
            ab_1.getBoundary(v3_1, v3_2);
            Vec3.min(v3_min, v3_min, v3_1);
            Vec3.max(v3_max, v3_max, v3_2);
        }
        if (this._modelBounds && this._worldBounds) {
            aabb.fromPoints(this._modelBounds, v3_min, v3_max);
            // @ts-ignore TS2445
            this._modelBounds.transform(root._mat, root._pos, root._rot, root._scale, this._worldBounds);
        }
    }

    public updateUBOs (stamp: number) {
        super.updateUBOs(stamp);
        for (let i = 0; i < this._joints.length; i++) {
            const { indices, buffers, transform, bindpose } = this._joints[i];
            Mat4.multiply(m4_1, transform.world, bindpose);
            for (let b = 0; b < buffers.length; b++) {
                uploadJointData(this._dataArray[buffers[b]], indices[b] * 12, m4_1, i === 0);
            }
        }
        for (let b = 0; b < this._buffers.length; b++) {
            this._buffers[b].update(this._dataArray[b]);
        }
        return true;
    }

    public initSubModel (idx: number, subMeshData: RenderingSubMesh, mat: Material) {
        const original = subMeshData.vertexBuffers;
        subMeshData.vertexBuffers = subMeshData.jointMappedBuffers;
        super.initSubModel(idx, subMeshData, mat);
        subMeshData.vertexBuffers = original;
    }

    public getMacroPatches (subModelIndex: number) : any {
        const superMacroPatches = super.getMacroPatches(subModelIndex);
        if (superMacroPatches) {
            return myPatches.concat(superMacroPatches);
        } else {
            return myPatches;
        }
    }

    public updateLocalBindings (psoci: IPSOCreateInfo, submodelIdx: number) {
        super.updateLocalBindings(psoci, submodelIdx);
        const bindingLayout = psoci.bindingLayout;
        const buffer = this._buffers[this._bufferIndices![submodelIdx]];
        if (buffer) { bindingLayout.bindBuffer(UBOSkinning.BLOCK.binding, buffer); }
    }

    private _ensureEnoughBuffers (count: number) {
        for (let i = 0; i < count; i++) {
            if (!this._buffers[i]) {
                this._buffers[i] = this._device.createBuffer({
                    usage: GFXBufferUsageBit.UNIFORM | GFXBufferUsageBit.TRANSFER_DST,
                    memUsage: GFXMemoryUsageBit.HOST | GFXMemoryUsageBit.DEVICE,
                    size: UBOSkinning.SIZE,
                    stride: UBOSkinning.SIZE,
                });
            }
            if (!this._dataArray[i]) {
                this._dataArray[i] = new Float32Array(UBOSkinning.COUNT);
            }
        }
    }
}

File: native/cocos/scene/SkinningModel.cpp
/****************************************************************************
 Copyright (c) 2021-2022 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
 worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You shall
 not use Cocos Creator software for developing other software or tools that's
 used for developing games. You are not granted to publish, distribute,
 sublicense, and/or sell copies of Cocos Creator.

 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
****************************************************************************/

#include "scene/SkinningModel.h"
#include <utility>
#include "scene/RenderScene.h"

namespace cc {
namespace scene {

std::vector<JointTransform*> SkinningModel::transStacks;

void SkinningModel::updateWorldMatrix(JointInfo* info, uint32_t stamp) {
    transStacks.clear();

    int i = -1;
    _worldMatrix.setIdentity();
    auto* currTransform = &info->transform;
    auto  parentSize    = static_cast<int>(info->parents.size());
    while (currTransform->node) {
        if ((currTransform->stamp == stamp || currTransform->stamp + 1 == stamp) && !currTransform->node->getFlagsChanged()) {
            _worldMatrix.set(currTransform->world);
            currTransform->stamp = stamp;
            break;
        }
        currTransform->stamp = stamp;
        transStacks.emplace_back(currTransform);
        i++;
        if (i >= parentSize) {
            break;
        }
        currTransform = &info->parents[i];
    }
    while (i > -1) {
        currTransform = transStacks[i--];
        auto* node    = currTransform->node;
        Mat4::fromRTS(node->getRotation(), node->getPosition(), node->getScale(), &currTransform->local);
        Mat4::multiply(_worldMatrix, currTransform->local, &currTransform->world);
        _worldMatrix.set(currTransform->world);
    }
}

void SkinningModel::updateUBOs(uint32_t stamp) {
    Model::updateUBOs(stamp);
    uint32_t bIdx = 0;
    Mat4     mat4;
    for (const JointInfo& jointInfo : _joints) {
        Mat4::multiply(jointInfo.transform.world, jointInfo.bindpose, &mat4);
        for (uint32_t buffer : jointInfo.buffers) {
            uploadJointData(jointInfo.indices[bIdx] * 12, mat4, _dataArray[buffer]);
            bIdx++;
        }
        bIdx = 0;
    }
    if (_realTimeTextureMode) {
        updateRealTimeJointTextureBuffer();
    } else {
        bIdx = 0;
        for (gfx::Buffer* buffer : _buffers) {
            buffer->update(_dataArray[bIdx], buffer->getSize());
            bIdx++;
        }
    }
}

void SkinningModel::uploadJointData(uint32_t base, const Mat4& mat, float* dst) {
    memcpy(reinterpret_cast<void*>(dst + base), mat.m, sizeof(float) * 12);
    dst[base + 3]  = mat.m[12];
    dst[base + 7]  = mat.m[13];
    dst[base + 11] = mat.m[14];
}

SkinningModel::~SkinningModel() {
    size_t count = _dataArray.size();
    for (size_t i = 0; i < count; i++) {
        if (_dataArray[i]) delete[] _dataArray[i];
    }
    _dataArray.clear();
    if (_realTimeJointTexture) {
        delete _realTimeJointTexture;
        _realTimeJointTexture = nullptr;
    }
}

void SkinningModel::setBuffers(std::vector<gfx::Buffer*> buffers) {
    _buffers = std::move(buffers);
    size_t count = _buffers.size();
    _dataArray.resize(_buffers.size());
    for (size_t i = 0; i < count; i++) {
        _dataArray[i]= new float[pipeline::UBOSkinning::count];
    }
}

void SkinningModel::updateTransform(uint32_t stamp) {
    auto* root = getTransform();
    if (root->getFlagsChanged() || root->getDirtyFlag()) {
        root->updateWorldTransform();
        _transformUpdated = true;
    }
    Vec3 v3Min{INFINITY, INFINITY, INFINITY};
    Vec3 v3Max{-INFINITY, -INFINITY, -INFINITY};
    AABB ab1;
    Vec3 v31;
    Vec3 v32;
    for (JointInfo& jointInfo : _joints) {
        updateWorldMatrix(&jointInfo, stamp);
        jointInfo.bound->transform(_worldMatrix, &ab1);
        ab1.getBoundary(&v31, &v32);
        Vec3::min(v3Min, v31, &v3Min);
        Vec3::max(v3Max, v32, &v3Max);
    }
    if (_modelBounds.getValid() && _worldBounds) {
        AABB::fromPoints(v3Min, v3Max, &_modelBounds);
        _modelBounds.transform(root->getNodeLayout()->worldMatrix, _worldBounds);
    }

    if (_scene) {
        _scene->updateOctree(this);
    }
}

void SkinningModel::setRealTimeJointTextures(std::vector<gfx::Texture *> textures) {
    if (textures.empty()) return;
    _realTimeTextureMode = true;
    _realTimeJointTexture = new RealTimeJointTexture();
    uint32_t length = 4 * RealTimeJointTexture::WIDTH * RealTimeJointTexture::HEIGHT;
    size_t count = _dataArray.size();
    for (size_t i = 0; i < count; i++) {
       delete[] _dataArray[i];
       _dataArray[i] = new float[length];
    }

    _realTimeJointTexture->textures = std::move(textures);
    _realTimeJointTexture->buffer = new float[length];
}

void SkinningModel::updateRealTimeJointTextureBuffer()
{
    uint32_t bIdx = 0;
    uint32_t width = RealTimeJointTexture::WIDTH;
    uint32_t height = RealTimeJointTexture::HEIGHT;
    for (gfx::Texture* texture: _realTimeJointTexture->textures) {
        auto *buffer = _realTimeJointTexture->buffer;
        auto *src    = _dataArray[bIdx];
        uint32_t count = width;
        uint32_t index0 = 0;
        uint32_t index1 = 0;
        for (uint32_t i = 0; i < count; i++) {
            index0 = 4 * i;
            buffer[index0++] = src[index1++];
            buffer[index0++] = src[index1++];
            buffer[index0++] = src[index1++];
            buffer[index0++] = src[index1++];
            index0 = 4 * (i + width);
            buffer[index0++] = src[index1++];
            buffer[index0++] = src[index1++];
            buffer[index0++] = src[index1++];
            buffer[index0++] = src[index1++];
            index0 = 4 * (i + 2 * width);
            buffer[index0++] = src[index1++];
            buffer[index0++] = src[index1++];
            buffer[index0++] = src[index1++];
            buffer[index0++] = src[index1++];
        }
        cc::gfx::TextureSubresLayers layer;
        cc::gfx::Offset texOffset;
        cc::gfx::Extent extent = {width, height, 1};
        cc::gfx::BufferTextureCopy region = {
           width,
           height,
           texOffset,
           extent,
           layer
        };
        auto *device = cc::gfx::Device::getInstance();
        device->copyBuffersToTexture(reinterpret_cast<const uint8_t *const *>(&buffer), texture, &region, 1);
        bIdx++;
    }
}

} // namespace scene
} // namespace cc
